File: app/src/main/java/com/stephenmorgandevelopment/thelinuxmanual/repos/UbuntuRepository.java
package com.stephenmorgandevelopment.thelinuxmanual.repos;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.util.Log;

import androidx.lifecycle.LiveData;

import com.stephenmorgandevelopment.thelinuxmanual.CommandSyncService;
import com.stephenmorgandevelopment.thelinuxmanual.data.DatabaseHelper;
import com.stephenmorgandevelopment.thelinuxmanual.data.LocalStorage;
import com.stephenmorgandevelopment.thelinuxmanual.distros.UbuntuHtmlApiConverter;
import com.stephenmorgandevelopment.thelinuxmanual.models.Command;
import com.stephenmorgandevelopment.thelinuxmanual.models.Release;
import com.stephenmorgandevelopment.thelinuxmanual.models.SimpleCommand;
import com.stephenmorgandevelopment.thelinuxmanual.network.HttpClient;
import com.stephenmorgandevelopment.thelinuxmanual.utils.Helpers;

import java.io.IOException;
import java.util.List;
import java.util.Map;

import io.reactivex.Single;
import io.reactivex.schedulers.Schedulers;
import okhttp3.Response;

public class UbuntuRepository implements ManPageRepository {
	private static UbuntuRepository instance;
	public static final String TAG = "UbuntuRepository";

	public static UbuntuRepository getInstance() {
		if (instance == null) {
			instance = new UbuntuRepository();
		}
		return instance;
	}

	private UbuntuRepository() {}

	public Single<Map<String, String>> getCommandData(SimpleCommand simpleCommand) {
		final LocalStorage storage = LocalStorage.getInstance();

		if (storage.hasCommand(simpleCommand.getId())) {
			try {
				return Single.just(storage.loadCommand(simpleCommand.getId()).getData());
			} catch (IOException ioe) {
				Log.i(TAG, "Unexpected file error loading - " + simpleCommand.getName() + ": " + ioe.getMessage());
			}
		}

		if (Helpers.hasInternet()) {
			return fetchCommandData(simpleCommand.getUrl())
					.observeOn(Schedulers.io())
					.doAfterSuccess((dataMap) -> {
						storage.saveCommand(new Command(simpleCommand.getId(), dataMap));
					});
		}

		return Single.error(new Throwable("Must have internet."));
	}

	public Command getCommandFromStorage(SimpleCommand simpleCommand) {
		final LocalStorage storage = LocalStorage.getInstance();

		try {
			return storage.loadCommand(simpleCommand.getId());
		} catch (IOException ioe) {
			Log.i(TAG, "Unexpected file error loading - " + simpleCommand.getName() + ": " + ioe.getMessage());
		}

		return null;
	}

	public Single<Map<String, String>> fetchCommandData(String pageUrl) {
		return HttpClient.fetchCommandManPage(pageUrl)
				.subscribeOn(Schedulers.io())
				.observeOn(Schedulers.computation())
				.flatMap(UbuntuHtmlApiConverter::crawlForCommandInfo);
	}

	public synchronized Single<List<SimpleCommand>> getPartialMatches(String searchQuery) {
		return Single.just(
				DatabaseHelper.getInstance().partialMatches(searchQuery))
				.subscribeOn(Schedulers.io());
	}

	public Single<SimpleCommand> addDescription(SimpleCommand match) {
		return HttpClient.fetchDescription(match)
				.subscribeOn(Schedulers.io())
				.observeOn(Schedulers.computation())
				.flatMap(response -> mapResponseToSimpleCommand(response, match))
				.observeOn(Schedulers.io())
				.doAfterSuccess(simpleCommand ->
						DatabaseHelper.getInstance().updateCommand(simpleCommand));
	}

	private Single<SimpleCommand> mapResponseToSimpleCommand(
				Response response, SimpleCommand match)
			throws IOException {

		//TODO or not
		//TODO Blob n the question???

		//TODO Blob = more effecient:
		return Single.just(
				match.setDescriptionReturnSimpleCommand(
						UbuntuHtmlApiConverter.crawlForDescription(response.body().string())));

		//TODO NonBlob = better readable:
//		String description = UbuntuHtmlAdapter.crawlForDescription(response.body().string());
//		match.setDescriptionReturnSimpleCommand(description);
//
//		return Single.just(match);
	}

	public LiveData<String> launchSyncService() {
		if (!CommandSyncService.isWorking()) {
			Intent intent = new Intent();
			intent.putExtra(CommandSyncService.DISTRO, UbuntuHtmlApiConverter.NAME);

			return CommandSyncService.enqueueWork(Helpers.getApplicationContext(), intent);
		}

		return CommandSyncService.getProgress();
	}
}

File: app/src/main/java/com/stephenmorgandevelopment/thelinuxmanual/repos/UbuntuRepository.kt
package com.stephenmorgandevelopment.thelinuxmanual.repos

import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.fromHtml
import androidx.work.OneTimeWorkRequestBuilder
import androidx.work.OutOfQuotaPolicy
import androidx.work.WorkManager
import com.stephenmorgandevelopment.thelinuxmanual.R
import com.stephenmorgandevelopment.thelinuxmanual.data.LocalStorage
import com.stephenmorgandevelopment.thelinuxmanual.data.SimpleCommandsDatabase
import com.stephenmorgandevelopment.thelinuxmanual.distros.ubuntu.UbuntuHtmlApiConverter
import com.stephenmorgandevelopment.thelinuxmanual.models.Command
import com.stephenmorgandevelopment.thelinuxmanual.models.MatchingItem
import com.stephenmorgandevelopment.thelinuxmanual.network.HttpClient
import com.stephenmorgandevelopment.thelinuxmanual.sync.CommandSyncWorker
import com.stephenmorgandevelopment.thelinuxmanual.utils.Helpers
import com.stephenmorgandevelopment.thelinuxmanual.utils.dlog
import com.stephenmorgandevelopment.thelinuxmanual.utils.ilog
import com.stephenmorgandevelopment.thelinuxmanual.utils.isNotNull
import com.stephenmorgandevelopment.thelinuxmanual.utils.stringFromRes
import com.stephenmorgandevelopment.thelinuxmanual.utils.wlog
import io.reactivex.schedulers.Schedulers
import kotlinx.coroutines.DelicateCoroutinesApi
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.launch
import kotlinx.coroutines.rx2.await
import kotlinx.coroutines.withContext
import java.io.IOException
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.min

@Singleton
class UbuntuRepository @Inject constructor(
    private val httpClient: HttpClient,
    private val roomDb: SimpleCommandsDatabase,
    private val localStorage: LocalStorage,
) {

    suspend fun getMatchingItemById(id: Long): MatchingItem? = withContext(Dispatchers.IO) {
        roomDb.dao().getCommandBy(id)
    }

    suspend fun getCommandData(simpleCommand: MatchingItem): Map<String, String> =
        withContext(Dispatchers.IO) {
            if (localStorage.hasCommand(simpleCommand.id)) {
                return@withContext getCommandFromStorage(simpleCommand.id)
            }

            if (Helpers.hasInternet()) {
                return@withContext fetchCommandData(simpleCommand.url)
                    .also { saveCommandInBackground(simpleCommand.id, it) }
            } else {
                return@withContext emptyMap<String, String>()
            }
        }

    suspend fun addDescription(matchedItem: MatchingItem): MatchingItem {
        return withContext(Dispatchers.Default) {
            httpClient.fetchDescription(matchedItem).await()
                ?.let {
                    UbuntuHtmlApiConverter.crawlForDescriptionAndSections(it.body?.string())
                }?.let { pair ->
                    val description = AnnotatedString.fromHtml(
                        pair.first ?: stringFromRes(R.string.no_description)
                    ).text.run { substring(0, min(length, 160)) }

                    val sections = pair.second.filterNotNull().toList()

                    matchedItem.copy(descriptionPreview = description, sections = sections)
                }?.also {
                    this.async(Dispatchers.IO) { roomDb.dao().update(it) }.start()
                } ?: matchedItem
        }
    }

    private val syncRequest
        get() = OneTimeWorkRequestBuilder<CommandSyncWorker>()
            .setExpedited(OutOfQuotaPolicy.RUN_AS_NON_EXPEDITED_WORK_REQUEST)
            .build().also {
                javaClass.ilog("Creating new instance of sync request.")
            }

    internal fun startSync(): Unit {
        WorkManager.getInstance(Helpers.getApplicationContext())
            .enqueue(syncRequest)
    }

    fun launchSyncWorker(): Flow<String> {
        if (!CommandSyncWorker.working) {
            startSync()
        }

        return CommandSyncWorker.progress
    }

    @OptIn(DelicateCoroutinesApi::class)
    private fun saveCommandInBackground(id: Long, data: Map<String, String>) {
        GlobalScope.launch {
            withContext(Dispatchers.IO) {
                localStorage.saveCommand(
                    Command(id, data)
                )
            }
        }.invokeOnCompletion { e ->
            if (e.isNotNull()) javaClass.dlog("Encountered error saving $id - ${e?.message}")
        }
    }

    private suspend fun getCommandFromStorage(id: Long): Map<String, String> {
        return try {
            withContext(Dispatchers.IO) {
                localStorage.loadCommand(id).data
            }
        } catch (e: IOException) {
            javaClass.wlog("Unexpected file error loading - $id: $e")
            emptyMap<String, String>()
        }
    }

    private suspend fun fetchCommandData(pageUrl: String): Map<String, String> {
        return try {
            httpClient.fetchCommandManPage(pageUrl)
                .doOnError {}
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.computation())
                .flatMap(UbuntuHtmlApiConverter::crawlForCommandInfo)
                .await()
        } catch (e: Throwable) {
            javaClass.wlog("Error fetching command data: ${e.message}")
            emptyMap()
        }
    }
}
